archunit/src/main/java/com/tngtech/archunit/ArchConfiguration.java

```diff
@@ -311,13 +313,23 @@ void remove(String propertyName) {
         Properties getMergedProperties() {
             Properties result = createProperties(this.properties);
             Properties overwritten = getSubProperties("archunit", System.getProperties());
-            if (!overwritten.isEmpty()) {
-                LOG.info("Merging properties: The following properties have been overwritten by system properties: {}", overwritten);
+
+            if (overwrittenPropertiesHaveNotChanged(overwritten)) {
+                return result;
             }
+
+            LOG.info("Merging properties: The following properties have been overwritten by system properties: {}", overwritten);
+            this.overwrittenProperties = overwritten;
+
             result.putAll(overwritten);
+            this.properties = result;
```

Comment:
Why do you overwrite the original `properties`? Isn't that way the basic configuration gone afterwards?
I would expect the behavior
```
getProperty("foo") -> bar from archunit.properties
System.setProperty("archunit.foo", "baz")
getProperty("foo") -> baz from System properties
System.clearProperty("archunit.foo")
getProperty("foo") -> again bar from archunit.properties
```

That behavior would now be broken, wouldn't it? Because the initial properties will never be restored to the ones from the `archunit.properties` if system properties are unset?

Reply:
I did this because I misunderstood the structure of the code. After your explanation in one of these comments I see it does indeed not make sense